src/libsystemd-network/icmp6-packet.c

```diff
@@ -0,0 +1,115 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <netinet/icmp6.h>
+
+#include "icmp6-packet.h"
+#include "icmp6-util.h"
+#include "in-addr-util.h"
+#include "iovec-util.h"
+#include "network-common.h"
+#include "socket-util.h"
+
+DEFINE_TRIVIAL_REF_UNREF_FUNC(ICMP6Packet, icmp6_packet, mfree);
+
+static ICMP6Packet* icmp6_packet_new(size_t size) {
+        ICMP6Packet *p;
+
+        if (size > SIZE_MAX - offsetof(ICMP6Packet, raw_packet))
+                return NULL;
+
+        p = malloc0(offsetof(ICMP6Packet, raw_packet) + size);
+        if (!p)
+                return NULL;
+
+        p->raw_size = size;
+        p->n_ref = 1;
+
+        return p;
+}
+
+int icmp6_packet_get_sender_address(ICMP6Packet *p, struct in6_addr *ret) {
+        assert(p);
+
+        if (in6_addr_is_null(&p->sender_address))
+                return -ENODATA;
+
+        if (ret)
+                *ret = p->sender_address;
+        return 0;
+}
+
+int icmp6_packet_get_timestamp(ICMP6Packet *p, clockid_t clock, usec_t *ret) {
+        assert(p);
+        assert(ret);
+
+        if (!TRIPLE_TIMESTAMP_HAS_CLOCK(clock) || !clock_supported(clock))
+                return -EOPNOTSUPP;
+
+        if (!triple_timestamp_is_set(&p->timestamp))
+                return -ENODATA;
+
+        *ret = triple_timestamp_by_clock(&p->timestamp, clock);
+        return 0;
+}
+
+static const struct icmp6_hdr* icmp6_packet_get_header(ICMP6Packet *p) {
+        assert(p);
+        assert(p->raw_size >= sizeof(struct icmp6_hdr));
+        return (const struct icmp6_hdr*) p->raw_packet;
+}
+
+uint8_t icmp6_packet_get_type(ICMP6Packet *p) {
```

Comment:
i'd probably change this to return the type as an `int`, so that we can return errors here too. If the function returns < 0 we'd have an error, and >= 0 would be the type. Which is safe, given that UINT8_MAX < INT_MAX